elks/arch/i86/kernel/irq.c

```diff
@@ -72,6 +73,42 @@ static void default_handler(int i, void *regs, void *dev)
 	printk("Unexpected interrupt: %u\n", i);
 }
 
+
+typedef void (* int_proc) (void);
+
+struct int_handler {
+	byte_t call;  // CALLF opcode (9Ah)
+	int_proc proc;
+	word_t seg;
+	byte_t irq;
+} __attribute__ ((packed));
+
+typedef struct int_handler int_handler_s;
+
+// TODO: move to irq.h
+int int_vector_set (int vect, int_proc proc, int seg);
+void _irqit (void);
+
+// TODO: move to segment.h
+int seg_data (void);
+
+// Add a dynamically allocated handler
+// that redirects to the generic handler
+
+static int int_handler_add (int irq, int vect)
+	{
+	int_handler_s * h = (int_handler_s *) heap_alloc (sizeof (int_handler_s), HEAP_TAG_INTHAND);
+	if (!h) return -ENOMEM;
+
+	h->call = 0x9A;    // CALLF opcode
+	h->proc = _irqit;  // generic interrupt handler
+	h->irq  = irq;
+
+	h->seg = int_vector_set (vect, (int_proc) h, seg_data ());
```

Comment:
Don't use another function wrapper `seg_data()`. Use existing `kernel_ds`, as used all over elsewhere in kernel.
Consider why even pass argument at all, since almost all other kernel ASM uses existing DS for accessing kernel data.

Reply:
1. Global variable `kernel_ds` is today questionnable, as it resides in a code segment (moreover, the code segment is not the same for all the functions as you switched to somewhat a medium model). I already complained a lot about this point, and asked many times to move that variable in the well known INIT segment. So I prefer to wrap it to not be impacted by the move of this variable (one day ?), at the cost of a very few bytes.

2. You need this segment argument for installing your 'fast handlers', as such handlers are not on the heap so in the data segment, but in the code segment. But the question is : 'what code segment" ? Again because of the medium memory model...

Reply:
> * Global variable `kernel_ds` is today questionnable, as it resides in a code segment

No - you are confusing `kernel_ds`, which is used everywhere, and `ds_kernel`, which is used in irqtab.S.
Even if they were the same, I don't agree with changing only in one place and leaving everything else. That's how the kernel got so messy in the first place over 20 years.

Reply:
Oup's... yes, sorry, got confused... will replace `seg_data()` by `kernel_ds`. But still expecting `ds_kernel` to move to the INIT segment one day...

Reply:
Hey I'm not fighting you, just trying to keep everything straight and working over here! 😏 There's still tons to do to clean ELKS up, I agree with you there.

Reply:
>  But still expecting `ds_kernel` to move to the INIT segment one day...

You mean the "setup" INITSEG? I thought, and agree with, your comments from earlier today in https://github.com/jbruchon/elks/pull/938#pullrequestreview-660324248 that all INIT probing code (and data) should be removed eventually. That means almost no need for INITSEG, except for boot-parameter passing. I would like to see INITSEG go away, and my last PR #941 takes another step in that direction.

Reply:
The fact is that the INITSEG is the only fixed anchor in the memory layout, apart the BDA and the vector table, and able to write to... plus the fact that the init code installs (and may decompresses) the kernel segments in memory, so it knowns very well where is the kernel code segment. But any data that is not related to this memory installation is still to be wiped out...

Reply:
Done in bfdabb7

Reply:
Discussion about `ds_kernel` has no more matter, as that variable could be now merely dropped.